### ISSUE TITLE
Whitelist for Non-Detected Axes?

### ISSUE THREAD

**Issue by Darkosto** [gigaherz/Survivalist#8]
Hey Gigaherz!

Would you be interested in creating a whitelist in the config to assign mod axes that are not detected by Survivalist and be able to assign them tool level?

Thank you! :)
Darkosto

**Comment by gigaherz** [comment]
Hmf... It would certainly be possible, but... If they aren't detected by my mod, it means they aren't really axes, so far as Minecraft is concerned. Let me explain:

When the chopping block is clicked, I run the following code:
```heldItem.getItem().getHarvestLevel(heldItem, "axe")```
Which first checks if the tool has an axe harvest level registered within Forge, and if not, then it checks if the vanilla-style tool class is an axe, and if so, returns the tool material's harvest level (wood is 0, iron is 2, etc.).

When a mod adds an axe, they should either call `setHarvestLevel("axe", level)`, set the `toolClass` to "axe", or extend `ItemAxe`. So if they aren't doing that, you should first ask them, and if they don't respond, then we can look at alternatives...

**Comment by Darkosto** [comment]
That was the first place I checked, so I talked a bit with the mod author and he assures me it's registered as an axe. The modded axe I'm talking about is also shown as the correct harvest level in-game which is why I was a bit confused. Veinminer does the same thing and tends to miss an item here and there. I know how much mod devs love having to create code for other's mistakes :P

I'll double check again to make sure the axes I'm specifically talking about are registered correctly and make sure it's not on our end. Thank you!

**Comment by Funwayguy** [comment]
I believe using 'canHarvestBlock()' as a fallback would support more items not specifically considered axes but still able to chop wood. You could also use some math on 'getStrVsBlock()' in order to calculate a specific level out of them (clamped within acceptable bounds of course).

**Comment by gigaherz** [comment]
Thanks, both of those sound like good ideas. However, both of them require an IBlockState to calculate the result, and I do not currently have an actual block, only a single-item ItemStack. I wish there was some 1:1 way to map items to blocks, then a lot of things would be a lot easier. I guess I'll leave things like they are now, since I don't want to write some hacky code to try to figure out an IBlockState from the item, but I will keep those suggestions in mind.

**Comment by Funwayguy** [comment]
Fair enough but for future reference a small piece of code like this would suffice for getting the IBlockState:
`
if(stack == null || !(stack.getItem() instanceof ItemBlock))
    	{
    		return null;
    	}
    	
    	ItemBlock iBlock = (ItemBlock)stack.getItem();
    	return iBlock.getBlock().getDefaultState();
`

**Comment by gigaherz** [comment]
Sadly, it would not be enough. Many blocks have sub-blocks using blockstates, properties such as type or tier or such. I can't just assume that all the sub-blocks will have the same properties as the default state, and .getStateFromMeta does NOT necessarily take item metadata values (there's no equivalence between item meta, and block meta). Because of that, the best (and ugliest) way to obtain an IBlockState from an ItemStack, is to call onItemUse, with a fake world, and see what IBlockState the item tried to set into the world. And that's what I meant by I don't want to write hacky code.

**Comment by Funwayguy** [comment]
Actually I was going to suggest Block.onBlockPlaced() instead of getStateFromMeta but I see where you're coming from. Thanks for looking into this anyway

**Comment by DogtorColmillo** [comment]
Hi, I post a few days ago on Project Curse this problem. Mods like Biomes O' Plenty and Hardcore Survival adds  Mud and Flint tools... neither are detected by this mod... is a shame. I will hope you add a system like veinminer to add new tools for mining and chopping trees will be a nice solution... meanwhile, I need to come back to wooden tools, and add some recipes with CraftTweaker for do a bypass, because  make an axe or a pickaxe is close imposible (in a hardcore game)...

**Comment by gigaherz** [comment]
I don't understand why modders make tools without making them BE tools... I'll end up having to implement shitty code in my mod, just to bypass other people's shitty code... :/

**Comment by gigaherz** [comment]
I just released v1.7.0, which adds a config category where items can be registered as chopping block axes.

It looks like:
```
customaxes {
    I:"minecraft:flint@0"=1
    I:"minecraft:string@0"=2
}
```
(the @0 is optional)
